### PR TITLE
Create Books Table

### PR DESCRIPTION
**Pull request by TheRealJessicaLi**
## Notion ticket link

[Create Books Table](https://www.notion.so/uwblueprintexecs/Create-Books-Table-Model-9afa1df91cca4cf6983fcfd1be43a8b5)

## Implementation description
* Created models/migrations for books, authors, publishers, and series (with some intermediate tables)

## Steps to test
1. Verify that tables are created in the backend

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/typescript/migrations/20211016164130-create-books.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+export const up: Migration = async ({ context: sequelize }) => {
+  await sequelize.getQueryInterface().createTable("books", {
+    id: {
+      allowNull: false,
+      primaryKey: true,
+      type: DataType.INTEGER,
+    },
+    review_id: {
+      allowNull: false,
+      type: DataType.INTEGER,
```
[the remark this answers is not in the document]
Comment (TheRealJessicaLi) [reply backend/typescript/migrations/20211016164130-create-books.ts:14]: Done!

File: backend/typescript/migrations/20211016172701-create-publishers.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+      type: DataType.INTEGER,
+      primaryKey: true,
+    },
+    full_name: {
+      allowNull: false,
+      type: DataType.STRING,
+    },
+    publish_date: {
+      allowNull: false,
+      type: DataType.DATE,
```
[the remark this answers is not in the document]
Comment (TheRealJessicaLi) [reply backend/typescript/migrations/20211016172701-create-publishers.ts:18]: Done!

File: backend/typescript/migrations/20211016172702-create-book-authors.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import { DataType } from "sequelize-typescript";
+
+import { Migration } from "../umzug";
+
+export const up: Migration = async ({ context: sequelize }) => {
+  await sequelize.getQueryInterface().createTable("book_authors", {
```
[the remark this answers is not in the document]
Comment (TheRealJessicaLi) [reply backend/typescript/migrations/20211016172702-create-book-authors.ts:6]: Done!

File: backend/typescript/models/book.model.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  series_order?: string;
+
+  @Column({ type: DataType.ARRAY(DataType.STRING) })
+  illustrator?: string[];
+
+  @Column({ type: DataType.ARRAY(DataType.STRING) })
+  translator?: string[];
+
+  @Column({ type: DataType.JSON })
+  formats: DataType.JSON;
```
[the remark this answers is not in the document]
Comment (TheRealJessicaLi) [reply backend/typescript/models/book.model.ts:45]: Done!

File: backend/typescript/migrations/20211016164130-create-books.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+      allowNull: true,
+      type: DataType.ARRAY(DataType.STRING),
+    },
+    translator: {
+      allowNull: true,
+      type: DataType.ARRAY(DataType.STRING),
+    },
+    formats: {
+      allowNull: false,
+      type: DataType.JSON,
```
[the remark this answers is not in the document]
Comment (TheRealJessicaLi) [reply backend/typescript/migrations/20211016164130-create-books.ts:42]: Done!

File: backend/typescript/models/book.model.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+  @Column({ type: DataType.RANGE(DataType.INTEGER) })
+  age_range?: RangeDataType<IntegerDataType>;
+
+  @BelongsTo(() => Series)
+  series!: Series;
+
+  @BelongsToMany(() => Author, () => BookAuthor)
+  author!: Author[];
```
[the remark this answers is not in the document]
Comment (TheRealJessicaLi) [reply backend/typescript/models/book.model.ts:54]: Done!

File: backend/typescript/models/book_author.model.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+  Column,
+  DataType,
+  Model,
+  Table,
+  ForeignKey,
+} from "sequelize-typescript";
+import Author from "./author.model";
+import Book from "./book.model";
+
+@Table({ tableName: "book_authors" })
```
[the remark this answers is not in the document]
Comment (TheRealJessicaLi) [reply backend/typescript/models/book_author.model.ts:12]: Done!

File: backend/typescript/models/book_publisher.model.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  Column,
+  DataType,
+  Model,
+  Table,
+  ForeignKey,
+} from "sequelize-typescript";
+import Book from "./book.model";
+import Publisher from "./publisher.model";
+
+@Table({ tableName: "book_publishers" })
```
[the remark this answers is not in the document]
Comment (TheRealJessicaLi) [reply backend/typescript/models/book_publisher.model.ts:12]: Done!

File: backend/typescript/models/publisher.model.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+@Table({ tableName: "publishers" })
+export default class Publisher extends Model {
+  @Column({ type: DataType.INT })
+  id!: number;
+
+  @Column({ type: DataType.STRING })
+  full_name!: string;
+
+  @Column({ type: DataType.DATE })
```
[the remark this answers is not in the document]
Comment (TheRealJessicaLi) [reply backend/typescript/models/publisher.model.ts:20]: Done!

File: backend/typescript/models/author.model.ts [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+  Model,
+  Table,
+  BelongsToMany,
+} from "sequelize-typescript";
+import Book from "./book.model";
+import BookAuthor from "./book_author.model";
+
+@Table({ tableName: "authors" })
+export default class Author extends Model {
+  @Column({ type: DataType.INT })
```
[the remark this answers is not in the document]
Comment (TheRealJessicaLi) [reply backend/typescript/models/author.model.ts:15]: Ah I didn't know that, thanks!

File: backend/typescript/models/series.model.ts [hunk 11]
```diff
@@ -0,0 +1,15 @@
+/* eslint import/no-cycle: 0 */
+import { Column, DataType, Model, Table, HasMany } from "sequelize-typescript";
+import Book from "./books.model";
+
+@Table({ tableName: "series" })
+export default class Series extends Model {
+  @Column({ type: DataType.INT })
+  id!: number;
+
+  @Column({ type: DataType.STRING })
```
[the remark this answers is not in the document]
Comment (TheRealJessicaLi) [reply backend/typescript/models/series.model.ts:10]: Good catch, thanks!

File: backend/typescript/models/book.model.ts [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+  illustrator?: string[];
+
+  @Column({ type: DataType.ARRAY(DataType.STRING) })
+  translator?: string[];
+
+  @Column({ type: DataType.ARRAY(DataType.JSON) })
+  formats!: DataType.JSON[];
+
+  @Column({ type: DataType.RANGE(DataType.INTEGER) })
+  age_range?: RangeDataType<IntegerDataType>;
```
Comment (Tahmeed27) [inline backend/typescript/models/book.model.ts:45]:
```suggestion
  age_range!: RangeDataType<IntegerDataType>;
```

File: backend/typescript/models/book.model.ts [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
```
Comment (Tahmeed27) [inline backend/typescript/models/book.model.ts:46]: I think here we need to add a BelongsTo() to Review as well - sorry for not noticing this earlier, when you do this change please all add a HasMany() to Book in review.model.ts

**Review by Tahmeed27 — CHANGES_REQUESTED** [review]
Just one minor change faciliating the relationship between books and reviews in the model files

File: backend/typescript/models/book.model.ts (hunk 12 above)
Comment (TheRealJessicaLi) [reply backend/typescript/models/book.model.ts]: Done!

File: backend/typescript/models/book.model.ts (hunk 13 above)
Comment (TheRealJessicaLi) [reply backend/typescript/models/book.model.ts:46]: Done, thanks for pointing that out!

**Review by Tahmeed27 — APPROVED** [review]
LGTM

**Pull request merged**